kubemarine/procedures/backup.py

```diff
@@ -252,136 +257,387 @@ def export_kubernetes_version(cluster: KubernetesCluster) -> None:
+                 graceful_finish_barrier: int,
+                 ):
+        self.logger = logger
+        self.backup_directory = backup_directory
+        self.namespaced_resources = namespaced_resources
+        self.nonnamespaced_resources = nonnamespaced_resources
+        self.namespaced_resources_result_map: Dict[str, List[str]] = {}
+        self.nonnamespaced_resources_result: List[str] = []
+        self.closed = False
+        self.elapsed: float = 0
+        self.total_files = 0
+        self._graceful_finish_barrier = graceful_finish_barrier
+        self._task_queue: 'Queue[ParserPayload]' = Queue()
+        self._lock = threading.Lock()
+
+    def schedule(self, payload: ParserPayload) -> None:
+        with self._lock:
+            if not self.closed:
+                self._task_queue.put(payload)
+            else:
+                self._clear(payload)
+
+    def finish(self, graceful: bool) -> None:
+        with self._lock:
+            self._graceful_finish_barrier -= 1
+            if not self.closed and (self._graceful_finish_barrier == 0 or not graceful):
+                self._task_queue.put(ParserPayload("end", "", None))
+                self._graceful_finish_barrier = 0
+
+    def run(self) -> None:
+        try:
+            while True:
+                payload = self._task_queue.get()
+                if payload.event == 'end':
+                    return
+
+                start = time.time()
+                self._handle(payload)
+                # do not clear the currently processing payload in case of exception
+                self._clear(payload)
+                self.elapsed += (time.time() - start)
+        finally:
+            with self._lock:
+                self.closed = True
+                # Queue might be not empty in case of exception. Need to clear all pending payloads
+                for payload in self._task_queue.queue:
+                    self._clear(payload)
+
+    @staticmethod
+    def _clear(payload: ParserPayload) -> None:
+        if payload.event == 'do':
+            os.remove(payload.resource_path)
+
+    def _parse_identity(self, line: str, api_version: str, kind: str) -> Tuple[str, str]:
+        if kind == '':
+            kind = self._parse_single_line_property(line, 'kind')
+        if api_version == '':
+            api_version = self._parse_single_line_property(line, 'apiVersion')
+
+        return api_version, kind
+
+    @staticmethod
+    def _parse_single_line_property(line: str, key: str) -> str:
+        item_prop = line[2:]
+        val = ''
+        if item_prop.startswith(f'{key}:'):
+            val = yaml.safe_load(item_prop)[key]
+
+        return val
+
+    def _handle(self, payload: ParserPayload) -> None:
+        namespace = payload.namespace
+        if namespace:
+            self.logger.debug(f"Loading resources from namespace {namespace!r}...")
+        else:
+            self.logger.debug(f"Loading non-namespaced resources...")
```

Comment:
After running `export.kubernetes` task there are full mesh in logs on verbose level (because every thread writes it's own logs at the same time). Maybe we will add thread id to log messages to handle logs order on different threads?
Additionally, if some kubectl command was unsuccessful on some namespace, we can't understand, on which namespace it was. I think, additional logs are needed in case of exception.

Reply:
> Maybe we will add thread id to log messages to handle logs order on different threads?

Added

> if some kubectl command was unsuccessful on some namespace, we can't understand, on which namespace it was

There will be an error: `Unexpected exit, command: '... -n <namespace >...'`, but I added one more explicit logging.